### ISSUE TITLE
Windows support

### ISSUE THREAD

**Issue by nemequ** [quixdb/squash#86]
Squash currently doesn't work with Windows, but it shouldn't be too difficult to do so.  As far as getting it to compile is concerned, the main thing that needs to happen is for `mapped-file.c` to the [Windows Memory-Mapped Files API](https://msdn.microsoft.com/en-us/library/ms810613.aspx).  That is pretty easy.

The other component to this is figuring out how to work within the Windows ecosystem.  Where do we install plugins?  How do we build in Visual Studio?  Should we use some sort of package manager (NuGet?)?  How can we get CI set up (AppVeyor?  Visual Studio Online?)?  This is really the blocking issue for me, but someone already familiar with the Windows ecosystem could probably take care of this stuff in fairly short order…

**Comment by tomByrer** [comment]
Would be nice.
Alternately, a VM or container (eg Docker) image for squash would suffice.

**Comment by nemequ** [comment]
Just to be clear, this isn't something that is going to happen without assistance of a Windows developer.  I would probably be willing and able to write the code (the first paragraph in the original comment), but this isn't going to be resolved until someone steps up to take care of at least the other stuff.

I don't think a VM or container would be a worthwhile endeavor.  Squash is pretty [easy to build](https://github.com/quixdb/squash/blob/master/docs/building.md), and will be even easier once I release 1.0 and it starts getting into distributions (`apt-get install squash libsquash libsquash-dev`/`dnf install squash libsquash libsquash-devel`/etc.)

**Comment by nemequ** [comment]
Good news, everyone!  I managed to get Squash compiling for Windows using mingw.

It's still very buggy; I haven't tested on Windows, but in Wine some of the unit tests actually work.

* Gipfeli won't compile.  I've submitted [a trivial patch](https://github.com/nemequ/gipfeli/commit/26c15a2531c9e486d15853c646e9629a873e20a9), but for now you'll have to disable it (pass -DENABLE_GIPFELI=no to cmake).
* yalz77 crashes due to division by 0 (‽).
* There is no default path for plugins, you'll need to use the `SQUASH_PLUGINS` environment variable to pass a path to the directory containing them.  This is the "Where do we install plugins?" question from the initial comment.
* Splicing doesn't work (data corruption), which also means the command-line tool doesn't work since it's basically a command-line wrapper around a `squash_splice` call.
* Memory-mapping is disabled.  (Note: this just means an optimization is disabled, it's not required functionality)

I'll try to figure out how to get it cross-compiling in Ubuntu so I can set up a Travis build to make sure I don't inadvertently break things more than they are, but it would be nice to have a native toolchain tested (i.e., set up [AppVeyor](http://www.appveyor.com/). Obviously we can't run the unit tests yet, but just making sure everything compiles would be a good first step.

I'll try to come up with a plan in the next few days for how to handle packaging and the `SQUASH_PLUGINS` issue, but I don't think I'll be actually implementing it.   I also have no plans to work on debugging the Windows build; mingw + wine is just too cumbersome, and actually using Windows is even worse for me.

Basically, Windows support is very close; someone who knows what they're doing on Windows just needs to spend a little time with it.

**Comment by jibsen** [comment]
Great news, thank you for your work on this.

**Comment by jibsen** [comment]
I can confirm it is compiling here with mingw-w64. I had to compile Ragel first and put that on the path, and I made a few minor changes to install the dll files (and fixed one bug).

The changes are at https://github.com/jibsen/squash, please note that I did not check if the changes to install break anything on other systems.

Also worth noting that you have to use forward slashes in the path in `SQUASH_PLUGINS`.

**Comment by jibsen** [comment]
After some work I got it to compile with MSVC as well. At the moment this is a bit hackish, so I have not pushed it.

**Comment by nemequ** [comment]
> The changes are at https://github.com/jibsen/squash, please note that I did not check if the changes to install break anything on other systems.

If you file a PR, Travis will test on Linux and, as of now 6293bba8, OS X.

I've gone ahead and merged what you had in there so far.  The `secure_getenv` fix was obviously correct (ugh, very embarrassing bug).  The install location stuff is exactly the kind of stuff I really need help with…

> Also worth noting that you have to use forward slashes in the path in SQUASH_PLUGINS.

Are you sure about that?  As of 3d5141fe you should be able to escape backslashes (with another backslash).

> After some work I got it to compile with MSVC as well. At the moment this is a bit hackish, so I have not pushed it.

That's great news!  That should be much more comfortable for a lot of people, and it should make it possible to get Squash building on AppVeyor.  You might want to rebase your changes on latest master, since I just got OS X working on Travis; that will help make sure it doesn't break other platforms.

If you want to file a PR just to abuse Travis feel free, I'll hold off on merging until you say it's ready.  If you're working on something, you can just keep amending a commit and doing forced pushes to your repo until it works as expected—every time your repo changes Travis will run another round of tests.  It's a very handy if you don't have access to the platform in question, though because of how many configurations Squash is tested with you'll have to wait about 10 minutes after pushing to see if it worked.

**Comment by nemequ** [comment]
> I'll try to come up with a plan in the next few days for how to handle packaging and the SQUASH_PLUGINS issue

I think this has pretty much been resolved with 3d5141fe.  This lets people define the default search path at compile time (with a cmake variable), and the `SQUASH_PLUGINS` environment varialbe should work better on Windows now (semicolon-separated lists, escaping with backslashes, etc.).  It might be a good idea to add a `SQUASH_EXTRA_PLUGINS` environment variable (or something) which would supplement the default search path instead of replace it (like `SQUASH_PLUGINS` does).

I'd say this is resolved once:

* MSVC is supported
* Unit tests work
* All plugins work

From there, there are I think the next step is to talk to some people from the nuget and/or chocolatey communities about how to structure packages for them.  They may have some ideas about what would go in the default search path, at least for nuget builds.  That would actually put Windows support ahead of Linux, unless I put together RPM and DEB packages first (which I plan to do soon).

And, of course, there is still the matter of getting mapped files working, but that's just an optimization so I'm not going to worry about it too much.
